test/test_runner/base_test_runner.go

```diff
@@ -101,9 +101,23 @@ func (t *BaseTestRunner) SetAgentConfig(agentConfig AgentConfig) {
 	t.AgentConfig = agentConfig
 }
 
-func (t *BaseTestRunner) RunAgent(runner *TestRunner) (status.TestGroupResult, error) {
+func (t *TestRunner) Run() status.TestGroupResult {
+	testName := t.TestRunner.GetTestName()
+	log.Printf("Running %v", testName)
+	testGroupResult, err := t.RunAgent()
+	if err == nil {
+		testGroupResult = t.TestRunner.Validate()
+	}
+	if testGroupResult.GetStatus() != status.SUCCESSFUL {
+		log.Printf("%v test group failed due to %v", testName, err)
+	}
+
+	return testGroupResult
+}
+
+func (t *TestRunner) RunAgent() (status.TestGroupResult, error) {
```

Comment:
We want the behavior to be that we have a default RunAgent method for all test_runner, and then each test_runner can implement their own version if provided.

Is there a test run to show that this is working as intended?

Reply:
yeah, `RunAgent` gets executed by the default `Run` and each test_runner can override it as needed

Reply:
This was ran from my personal fork https://github.com/movence/amazon-cloudwatch-agent/actions/runs/5158969625/jobs/9293404442
```
null_resource.integration_test_run (remote-exec): === RUN   TestUserdata
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:09 Testing all EC2 plugins
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:09 Running UserdataTest
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:09 Starting agent using agent config file agent_configs/config.json
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:09 Copy File agent_configs/config.json to /opt/aws/amazon-cloudwatch-agent/bin/config.json
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:09 File agent_configs/config.json abs path /home/ec2-user/amazon-cloudwatch-agent-test/test/userdata/agent_configs/config.json
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:10 File : agent_configs/config.json copied to : /opt/aws/amazon-cloudwatch-agent/bin/config.json
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:10 Agent has started
null_resource.integration_test_run: Still creating... [3m20s elapsed]
null_resource.integration_test_run: Still creating... [3m30s elapsed]
null_resource.integration_test_run: Still creating... [3m40s elapsed]
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:40 Agent has been running for : 30s
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 Agent is stopped
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 Delete file /opt/aws/amazon-cloudwatch-agent/bin/config.json
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 Removed file: /opt/aws/amazon-cloudwatch-agent/bin/config.json
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 instruction {InstanceId {<nil>}} provider HostDimensionProvider returned dimension {<nil> <nil> {}}
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 instruction {InstanceId {<nil>}} provider LocalInstanceIdDimensionProvider returned dimension {0xc000038220 0xc000038230 {}}
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 Result dim is : InstanceId, i-0544dd42c6d9882a7
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 instruction {cpu {0xc000159c30}} provider HostDimensionProvider returned dimension {<nil> <nil> {}}
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 instruction {cpu {0xc000159c30}} provider LocalInstanceIdDimensionProvider returned dimension {<nil> <nil> {}}
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 instruction {cpu {0xc000159c30}} provider LocalImageIdDimensionProvider returned dimension {<nil> <nil> {}}
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 instruction {cpu {0xc000159c30}} provider LocalInstanceTypeDimensionProvider returned dimension {<nil> <nil> {}}
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 instruction {cpu {0xc000159c30}} provider CustomDimensionProvider returned dimension {0xc0000382b0 0xc0000382c0 {}}
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 Result dim is : cpu, cpu-total
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 Metric query input dimensions
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 	Dimensions:
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 		Dim(name="InstanceId", val="i-0544dd42c6d[988](https://github.com/movence/amazon-cloudwatch-agent/actions/runs/5158969625/jobs/9293404442#step:7:989)2a7"
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 		Dim(name="cpu", val="cpu-total"
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 Metric data input: namespace UserdataTest, name cpu_time_active_userdata, stat Average, period 10
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 Metric values are : [361.73 361.08]
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 metric values are [361.73 361.08]
null_resource.integration_test_run (remote-exec): 2023/06/02 19:29:41 The average value 361.405000 for metric cpu_time_active_userdata are within bound [0.000000, 0.000000]
null_resource.integration_test_run (remote-exec): --- PASS: TestUserdata (31.34s)
```

Reply:
https://github.com/movence/amazon-cloudwatch-agent/actions/runs/5158969625/jobs/9293404442#step:7:970
I don't think it is. Userdata test implements `RunAgent` method so that it does nothing about starting and configuring the agent since those should be done using EC2 userdata. In your test run it's running the default RunAgent method.

Reply:
You are right. I should have overlooked the logs. I changed the userdata to have its own `Run` method, so it calls its own function not the one in `base_test_runner`.

```
null_resource.integration_test (remote-exec): === RUN   TestUserdata
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 Testing all EC2 plugins
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 Running UserdataTest
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 Running test for runAgent in userdata mode
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 instruction {InstanceId {<nil>}} provider HostDimensionProvider returned dimension {<nil> <nil> {}}
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 instruction {InstanceId {<nil>}} provider LocalInstanceIdDimensionProvider returned dimension {0xc000270070 0xc000270080 {}}
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 Result dim is : InstanceId, i-071bc4321aa09406d
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 instruction {cpu {0xc0001d3b20}} provider HostDimensionProvider returned dimension {<nil> <nil> {}}
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 instruction {cpu {0xc0001d3b20}} provider LocalInstanceIdDimensionProvider returned dimension {<nil> <nil> {}}
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 instruction {cpu {0xc0001d3b20}} provider LocalImageIdDimensionProvider returned dimension {<nil> <nil> {}}
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 instruction {cpu {0xc0001d3b20}} provider LocalInstanceTypeDimensionProvider returned dimension {<nil> <nil> {}}
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 instruction {cpu {0xc0001d3b20}} provider CustomDimensionProvider returned dimension {0xc000270100 0xc000270110 {}}
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 Result dim is : cpu, cpu-total
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 Metric query input dimensions
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 	Dimensions:
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 		Dim(name="InstanceId", val="i-071bc[432](https://github.com/movence/amazon-cloudwatch-agent/actions/runs/5164785471/jobs/9303925508#step:6:433)1aa09406d"
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 		Dim(name="cpu", val="cpu-total"
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 Metric data input: namespace UserdataTest, name cpu_time_active_userdata, stat Average, period 10
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 Metric values are : [125.72000000000003 110.20000000000002]
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 metric values are [125.72000000000003 110.20000000000002]
null_resource.integration_test (remote-exec): 2023/06/03 17:20:23 The average value 117.960000 for metric cpu_time_active_userdata are within bound [0.000000, 0.000000]
null_resource.integration_test (remote-exec): --- PASS: TestUserdata (0.06s)
```